### DISCUSSION TITLE
Show description before the type

### DISCUSSION THREAD

**Discussion by slinkydeveloper** [jy95/docusaurus-json-schema-plugin#80] [Ideas]
Hi,

I was wondering whether it's possible to show the description above the type, which is particularly useful for objects with a lot of fields.

**Comment by slinkydeveloper** [comment]
Perhaps at least for objects and arrays

**Comment by slinkydeveloper** [comment]
I tried to swizzle the component `CreateObject` but it didn't work, not sure why :(

**Comment by jy95** [comment]
@slinkydeveloper Welcome ;)

As you might have seen in the [documentation](https://jy95.github.io/docusaurus-json-schema-plugin/docs/api/JSONSchemaViewer), some customizations to fit everyone needs were done : 

- `qualifierMessagesOrder` for customization of qualifier message (e.g. you want "const" before "pattern")
- `DescriptionComponent` for customization of the rendering of the description itself (e.g. you use Markdown in your description)

Do you have an idea on how you would like to specify the control you want ? It could be interesting to deep the idea a bit more ...
> at least for objects and arrays

**Reply by slinkydeveloper** [reply]
> DescriptionComponent 

Side note about this: It would be cool to have some docs on how to render descriptions as markdown. I found it somewhere in the tests around this repo :)

> qualifierMessagesOrder

My understanding from the docs (and also reading the code) is that the `description` has a fixed location in the code, so this option won't have any effect.

I think what would be useful is to be able to swizzle only the `CreateObject` and `CreateArray` for me, and then change the order as I wish. From:

```
export default function CreateObject(props: Props): JSX.Element {
  const { schema, nullable, description } = props
  const options = useJSVOptionsContext()

  return (
    <>
      <TypeLabel />
      &nbsp;&#58;&nbsp;
      <ObjectLabel />
      <UnlistedProperties schema={schema} />
      <Properties schema={schema} />
      <PatternProperties schema={schema} />
      <PropertyNames schema={schema} />
      <AdditionalProperties schema={schema} />
      <UnevaluatedProperties schema={schema} />
      <div style={{ marginTop: "var(--ifm-table-cell-padding)" }}>
        <QualifierMessages
          schema={schema}
          options={options}
          nullable={nullable}
        />
      </div>
      {description !== undefined && (
        <CreateDescription description={description} />
      )}
    </>
  )
}
```

To: 

```
export default function CreateObject(props: Props): JSX.Element {
  const { schema, nullable, description } = props
  const options = useJSVOptionsContext()

  return (
    <>
      <TypeLabel />
      &nbsp;&#58;&nbsp;
      <ObjectLabel />
      {description !== undefined && (
        <CreateDescription description={description} />
      )}
      <UnlistedProperties schema={schema} />
      <Properties schema={schema} />
      <PatternProperties schema={schema} />
      <PropertyNames schema={schema} />
      <AdditionalProperties schema={schema} />
      <UnevaluatedProperties schema={schema} />
      <div style={{ marginTop: "var(--ifm-table-cell-padding)" }}>
        <QualifierMessages
          schema={schema}
          options={options}
          nullable={nullable}
        />
      </div>
    </>
  )
}
```

I managed to workaround this issue by simply swizzling the whole `JsonSchemaViewer` component, for some reason swizzling only `CreateObject` is not enough... Perhaps a good solution to this issue is to figure out why swizzling single components doesn't work and fix it?

**Reply by jy95** [reply]
> DescriptionComponent : Side note about this: It would be cool to have some docs on how to render descriptions as markdown. I found it somewhere in the tests around this repo :)

Nothing prevents someone to add an example in the  [docs](https://jy95.github.io/docusaurus-json-schema-plugin/docs/demo/viewer) ;)

> for some reason swizzling only CreateObject is not enough

Technically speaking, I followed what [Docusaurus](https://docusaurus.io/docs/api/plugin-methods) said in my [index.ts](https://github.com/jy95/docusaurus-json-schema-plugin/blob/main/src/index.ts#L89)

So I'm curious : which commands did you use for swizzling ? Something similar to the [documentation ](https://docusaurus.io/docs/typescript-support#swizzling-typescript-theme-components) ?

```bash
npm run swizzle docusaurus-json-schema-plugin JSONSchemaViewer -- --typescript
```

Why I ask is because [Docusaurus](https://docusaurus.io/docs/swizzling) actually recommend this : 
![image](https://github.com/jy95/docusaurus-json-schema-plugin/assets/9306961/bc392d9e-419e-4923-a18a-a23151b680e6) 

More info on https://docusaurus.io/docs/swizzling#wrapping

You can ask on Docusaurus discord / Github  what goes wrong (unlikely that this lib is the only one with nested components people would like to swizzle). If I have to update something, it will be with pleasure.

**Reply by slinkydeveloper** [reply]
I think wrapping won't solve my case anyway, as i need to change the order of the components, right? 

I tried the following command:

```
yarn swizzle docusaurus-json-schema-plugin JSONSchemaViewer/JSONSchemaElements/object/CreateObject --typescript --eject
```

This copied only the single file `CreateObject.tsx`. You can see the list of components that i can swizzle here:

```
yarn swizzle docusaurus-json-schema-plugin --list
```

I have only one hypothesis in mind: looking at the various swizzle examples I see there are no relative imports. Perhaps that's the trick that allows swizzling, some magical import resolution swapping? 

(BTW sorry if I ask noob questions, it's my first time with docusaurus and typescript/mdx/etc :smile:)

> You can ask on Docusaurus discord / Github what goes wrong (unlikely that this lib is the only one with nested components people would like to swizzle). If I have to update something, it will be with pleasure.

Will try to do that!

**Reply by jy95** [reply]
@slinkydeveloper I think that in [Docusaurus documentation](https://docusaurus.io/docs/api/plugin-methods) , there is something missing on how  to make one theme / plugin perfect for swizzle ...

What I would love to have in the current implementation is what you said : [Path aliases in Typescrript.](https://www.typescriptlang.org/tsconfig#paths)

From : 

```ts
import { User } from '../../user/model';
import { Article } from '../../article/model';

import { Cache } from '../../../../cache';
import { MongoDB } from '../../../../mongodb';
```

To :

```ts
import { User } from '@modules/user/model';
import { Article } from '@modules/article/model';

import { Cache } from '@services/cache';
import { MongoDB } from '@services/mongodb';
```

Unless I am wrong, with current Docusaurus V2 implementation, I cannot do that otherwise that an annoying long paths in my imports

```ts
import { User } from 'docusaurus-json-schema-plugin/JSONSchemaViewer/JSONSchemaElements/object/CreateObject;
import { Article } from 'docusaurus-json-schema-plugin/.../nestedPath/Component';
```

**Reply by jy95** [reply]
@slinkydeveloper Good news - one hellish  refactoring of imports / [tests](https://github.com/jy95/docusaurus-json-schema-plugin/pull/82) later, swizzling of components (like `CreateObject` / `CreateArray`) should work as expected - without any breaking changes for current users of the lib.
docusaurus-json-schema-plugin [1.6.1](https://www.npmjs.com/package/docusaurus-json-schema-plugin/v/1.6.1) is out with that 🎉

![image](https://github.com/jy95/docusaurus-json-schema-plugin/assets/9306961/bf2677c0-1121-4d01-826a-2f7b9d484177)

```tsx
import React from "react"

import Items from "@theme/JSONSchemaViewer/JSONSchemaElements/array/Items"
import Contains from "@theme/JSONSchemaViewer/JSONSchemaElements/array/Contains"
import PrefixItems from "@theme/JSONSchemaViewer/JSONSchemaElements/array/PrefixItems"
import AdditionalItems from "@theme/JSONSchemaViewer/JSONSchemaElements/array/AdditionalItems"
import UnevaluatedItems from "@theme/JSONSchemaViewer/JSONSchemaElements/array/UnevaluatedItems"

import { QualifierMessages } from "@theme/JSONSchemaViewer/utils"

import { useJSVOptionsContext } from "@theme/JSONSchemaViewer/contexts"

import { ArrayLabel, TypeLabel } from "@theme/JSONSchemaViewer/labels"

import { CreateDescription } from "@theme/JSONSchemaViewer/JSONSchemaElements"

import type { JSONSchemaNS } from "@theme/JSONSchemaViewer/types"

type Props = {
  schema: JSONSchemaNS.Array
  nullable?: boolean
  description?: string
  [x: string]: any
}

export default function CreateArray(props: Props): JSX.Element {
  const { schema, nullable, description } = props
  const options = useJSVOptionsContext()

  return (
    <>
      <TypeLabel />
      &nbsp;&#58;&nbsp;
      <ArrayLabel />
      {description !== undefined && (
        <CreateDescription description={description} />
      )}
      <PrefixItems schema={schema} />
      <Items schema={schema} />
      <AdditionalItems schema={schema} />
      <UnevaluatedItems schema={schema} />
      <Contains schema={schema} />
      <div style={{ marginTop: "var(--ifm-table-cell-padding)" }}>
        <QualifierMessages
          schema={schema}
          options={options}
          nullable={nullable}
        />
      </div>
    </>
  )
}
```

Your IDE might complain but it works ... Maybe later I can add something in https://github.com/jy95/docusaurus-json-schema-plugin/blob/main/src/theme-json-schema.d.ts for that ...

**Reply by slinkydeveloper** [reply]
This is great, it works like a charm! Thanks for that!
